File: haskell/sudoku.hs
-- sudoku solver using backtrack
module Main where

import Control.DeepSeq(deepseq)
import Data.Char(ord)
import Data.List(intercalate, elemIndex, find)
import Data.Time.Clock(diffUTCTime, getCurrentTime)
import System.Environment(getArgs)
import System.IO(hFlush, stdout)
import Text.Printf(printf)
import System.FilePath

boardSize = 9

data Board = Board [[Int]]

instance Show Board where
        show (Board b) = unlines $ map (unwords . map show) b

instance Read Board where
        readsPrec _ s = [(Board $ fromStr s, "")]

groupN _ [] = []
groupN n xs = a : groupN n b where
                    (a,b) = splitAt n xs

fromStr = groupN boardSize . map toInt
        where toInt = abs . (48 -) . ord

-- return the next row, column where value is zero
nextEmpty :: Board -> [Int]
nextEmpty (Board b) = case (elemIndex 0 (concat b)) of
        Just n  -> [n `div` boardSize, n `mod` boardSize]
        Nothing -> []

-- return true if val can be put at row and col of sudoku
canPut :: Board -> Int -> Int -> Int -> Bool
canPut (Board b) row col val =
        and [val `notElem` (b !! row), -- test row
                 val `notElem` [r !! col | r <- b], -- test column
        -- test square
             val `notElem` [b!!i!!j | i <- [sqX .. sqX+2], j <- [sqY .. sqY+2]]]
        where
                sqX = row - (row `mod` 3)
                sqY = col - (col `mod` 3)

updateList xs pos val = x ++ val : ys
        where (x,_:ys) = splitAt pos xs

updateSudoku :: Board -> Int -> Int -> Int -> Board
updateSudoku (Board b) row col val = Board $ updateList b row (updateList (b!!row) col val)

-- solve a sudoku, return the solved sudoku if found solution or the same otherwise
solve :: Board -> Board        
solve board
        | spot == [] = board -- sudoku is solved
        | otherwise =
                let (row:col:_) = spot
                in
                        case find (\s -> nextEmpty s == []) (map (\val->(if canPut board row col val
                                then solve (updateSudoku board row col val)
                                else board)) [1..9]) of
                                        Just s -> s
                                        Nothing -> board
        where
                spot = nextEmpty board

-- progress bar
loadBar :: Int -> Int -> Int -> Int -> IO()
loadBar step totalSteps resolution width =
        if (and [(div totalSteps resolution /= 0), (mod step (div totalSteps resolution) /= 0)])
                then do
                        let ratio = (fromIntegral step::Float) / (fromIntegral totalSteps::Float)
                        let count = floor (ratio * (fromIntegral width::Float))
                        let perc = floor (ratio * 100)
                        printf "%3d%% [" (perc::Int)
                        mapM (\_ -> printf "=") [0..count-1]
                        mapM (\_ -> printf " ") [count..width-1]
                        printf "]\r"
                        hFlush stdout
                        return()
                else
                        return()

process :: Int -> Int -> String -> IO String
process step totalSteps sudoku = do
        let board = (read sudoku)::Board
        let result = (show . solve) board
        result `deepseq` loadBar step totalSteps 20 50 -- deepseq forces evaluation of result
        return result

main = do
        -- read entire file into memory
        (input:_) <- getArgs
        contents <- readFile input
        let sudokus = lines contents
        let total = length sudokus

        -- solve all sudokus
        start <- getCurrentTime
        solutions <- mapM (\(step,sudoku) -> process step total sudoku) (zip [1..] sudokus)
        let result = unlines solutions
        end <- getCurrentTime

        -- present results
        putStrLn $ " -- Elapsed time: " ++ (show (diffUTCTime end start))
        writeFile ("solved_" ++ (takeFileName input)) result

File: javascript/sudoku.js
// Solve a sudoku using backtrack
const assert = require("assert");

const ROW_SIZE = 9;

class Sudoku {
  #board;

  constructor(text) {
    this.#board = new Array();
    for (let k = 0; k < ROW_SIZE; k++) {
      this.#board[k] = new Array();
    }
    let i = 0;
    let j = 0;
    for (let k = 0; k < text.length; k++) {
      if (!isNaN(text.charAt(k)) && i < ROW_SIZE && j < ROW_SIZE) {
        this.#board[i][j] = parseInt(text.charAt(k));
        i = j == ROW_SIZE - 1 ? i + 1 : i;
        j = (j + 1) % ROW_SIZE;
      }
    }
  }

  toString() {
    const result = new Array();
    for (let i = 0; i < this.#board.length; i++) {
      result[i] = this.#board[i].join(" ");
    }
    return result.join("\n");
  }

  nextEmpty() {
    for (let i = 0; i < this.#board.length; i++) {
      for (let j = 0; j < this.#board.length; j++) {
        if (this.#board[i][j] == 0) {
          return [i, j];
        }
      }
    }
    return null;
  }

  canPut(x, y, val) {
    for (let i = 0; i < ROW_SIZE; i++) {
      if (this.#board[x][i] == val) return false; // row test
      if (this.#board[i][y] == val) return false; // column test
    }
    // square test
    let sqX = x - (x % 3);
    let sqY = y - (y % 3);
    for (let i = sqX; i < sqX + 3; i++) {
      for (let j = sqY; j < sqY + 3; j++) {
        if (this.#board[i][j] == val) return false;
      }
    }
    return true;
  }

  solve() {
    let spot = this.nextEmpty();
    if (spot) {
      let x = spot[0];
      let y = spot[1];
      for (let val = 1; val < 10; val++) {
        if (this.canPut(x, y, val)) {
          this.#board[x][y] = val;
          this.solve();
          if (!this.nextEmpty()) {
            // solution found
            return;
          }
        }
      }
      // solution not found, backtracking
      this.#board[x][y] = 0;
    }
  }
}

// tests
(function () {
  const sudoku = new Sudoku(
    "200000060000075030048090100000300000300010009000008000001020570080730000090000004"
  );
  assert.deepEqual(sudoku.nextEmpty(), [0, 1]);
  assert.ok(sudoku.canPut(0, 1, 1));
  sudoku.solve();
  assert.ok(!sudoku.nextEmpty());
})();

// progress bar
var util = require("util");
function loadBar(step, totalSteps, resolution, width) {
  if (
    parseInt(totalSteps / resolution) != 0 &&
    step % parseInt(totalSteps / resolution) == 0
  ) {
    ratio = step / totalSteps;
    count = parseInt(ratio * width);
    process.stdout.write(util.format("%d%% [", parseInt(ratio * 100)));
    for (var i = 0; i < count; i++) {
      process.stdout.write("=");
    }
    for (var i = count + 1; i < width; i++) {
      process.stdout.write(" ");
    }
    process.stdout.write("]\r");
  }
}

// read entire sudokus into memory
const fs = require("fs");
const path = require("path");
const input = process.argv[2];
if (!input) {
  console.log("Input file necessary");
  return;
}
fs.readFile(input, "ascii", function (err, data) {
  if (err) {
    console.log(err);
    return;
  }
  let before = new Date().getTime();

  let result = "";

  // solve all sudokus
  let lines = data.replace(/^\s+|\s+$/g, "").split("\n");
  let total = lines.length;
  for (let i = 0; i < total; i++) {
    let sudoku = new Sudoku(lines[i]);
    sudoku.solve();
    result = result + sudoku.toString() + "\n\n";
    loadBar(i, total, 20, 50);
  }

  let after = new Date().getTime();

  // write solution to file
  let output = path.basename(input);
  fs.writeFile("solved_" + output, result, function (err) {
    if (!err) {
      console.log(" --Elapsed time: %d ms", after - before);
    } else {
      console.log(`Error: ${err}`);
    }
  });
});
